pkg/api/service/scans.go

```diff
@@ -339,64 +354,49 @@ func (s ScansService) ProcessScanCheckNotification(ctx context.Context, msg []by
 		return nil
 	}
 
-	// Don't take into account inconsistent progress in a message with a terminal status.
-	if checkStates.IsTerminal(checkMssg.Status) && (checkProgress != 1.0) {
-		_ = level.Error(s.logger).Log("FixingInvalidProgressInTerminalStatus", checkProgress, "Status", checkMssg.Status)
-		checkProgress = 1
-		checkMssg.Progress = &checkProgress
+	_, err = s.db.UpsertCheck(scanID, checkID, checkMssg, api.CheckStates.LessOrEqual(checkMssg.Status))
+	if err != nil {
+		return err
+	}
+	// As a check message does not contain all the information
+	// of a check we must merge with the the info of the check in the DB.
+	check := mergeChecks(dbCheck, checkMssg)
+	if err != nil {
+		return err
+	}
+	programID := ""
+	if check.Tag != nil {
+		programID = *check.Tag
 	}
+	s.notifyCheck(check, programID)
 
-	// The progress could still be incorrect if the check is not in a terminal status.
-	// In that case we want to discard the message because we can not deduce the progress
-	// from the status.
-	if checkProgress > 1.0 || checkProgress < 0.0 {
-		_ = level.Error(s.logger).Log("NotValidProgress", checkMssg.Progress)
+	// If the status of the check is not terminal it will not affect the status
+	// of the scan, so we are done.
+	if !api.CheckStates.IsTerminal(checkMssg.Status) {
 		return nil
 	}
 
-	// Only update scan and checks data if there is an status change.
-	// Intermediate check messages data (e.g.: progress reporting) are not persisted.
-	// We have to take into account the particular check message sent by the agent for
-	// report and raw data which does not state an explicit status.
-	if checkMssg.Status == "" || checkStates.IsHigher(checkMssg.Status, dbCheck.Status) {
-		checkCount, err := s.db.UpsertCheck(scanID, checkID, checkMssg, checkStates.LessOrEqual(checkMssg.Status))
-		if err != nil {
-			return err
-		}
-		if checkCount == 0 {
-			_ = level.Info(s.logger).Log("NoEffectProcessingCheckUpdate", string(msg))
-		}
-
-		scanCount, scan, err := s.updateScanStatus(scanID)
+	// Count the check as finished in its scan. Note that this operation is
+	// idempotent, that means: even if called multiple times, for a given check
+	// it will only increase by one the number of checks finished in the scan.
+	_, err = s.db.AddCheckAsFinished(checkID)
+	if err != nil {
+		return err
+	}
+	scanCount, status, err := s.updateScanStatus(scanID)
+	if err != nil {
+		return err
+	}
+	if scanCount > 0 {
+		_ = level.Info(s.logger).Log("ScanStatusUpdated", string(msg))
+		_ = level.Debug(s.logger).Log("ScanStatusSet", scanID.String()+";"+status)
+	}
+	if status == ScanStatusFinished {
+		err = s.notifyScan(scanID)
```

Comment:
Here we have the same problem as with the check metrics I believe.
Initially I was thinking that we should move this `if` and the `notifyScan` into the `if scanCount > 0` if, so we only push scan events and metrics if a status change has been updated in DB and it is a `FINISHED` status. But in the edge case that the "process" stops after the update in DB (and we have not pushed the metrics nor scan data and the message is not deleted from the queue) once the message is processed again, then we would not push the metrics because the "second" processing would not imply a change in DB...

This is tricky, let me think a little bit about it.